File: 3211-generate-binary-strings-without-adjacent-zeros/3211-generate-binary-strings-without-adjacent-zeros.cpp
class Solution {
public:
    void solve(int n,string &output,vector<string>&ans,int ind){
        if(ind>=n){
            ans.push_back(output);
            return;
        }
        if(output.empty() || output.back()!='0'){
            output.push_back('0');
            solve(n,output,ans,ind+1);
            output.pop_back();
        }
        output.push_back('1');
        solve(n,output,ans,ind+1);
        output.pop_back();

    }

    vector<string> validStrings(int n) {
        string output="";
        vector<string>ans;
        solve(n,output,ans,0);
        return ans;
    }
};

File: 3211-generate-binary-strings-without-adjacent-zeros/3211-generate-binary-strings-without-adjacent-zeros.java
class Solution {
   public void solve(int n,List<String>ans,StringBuilder output,int ind){
      if(ind>=n){
        ans.add(output.toString());
        return;
      }
      if(output.isEmpty() || output.charAt(output.length()-1)!='0'){
         output.append('0');
         solve(n,ans,output,ind+1);
         output.deleteCharAt(output.length()-1);
      }
      output.append('1');
      solve(n,ans,output,ind+1);
      output.deleteCharAt(output.length()-1);
   }

    public List<String> validStrings(int n) {
        List<String> ans=new ArrayList<>();
        StringBuilder output=new StringBuilder();
        solve(n,ans,output,0);
        return ans;
    }
}